src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -2147,6 +2180,48 @@ - (void)_removeExpectedValueForAttributePath:(MTRAttributePath *)attributePath e
     }
 }
 
+// This method checks if there is a need to inform delegate that the attribute cache has been "primed"
+//   - The delegate callback is only called once
+- (void)_checkIfCacheIsPrimed
+{
+    os_unfair_lock_assert_owner(&self->_lock);
+
+    // Only send the callback once per lifetime of MTRDevice
+    if (_delegateDeviceCachePrimedCalled) {
+        return;
+    }
+
+    // Check if root node descriptor exists
+    NSDictionary * rootDescriptorPartsListDataValue = _readCache[[MTRAttributePath attributePathWithEndpointID:@(0) clusterID:@(MTRClusterIDTypeDescriptorID) attributeID:@(MTRAttributeIDTypeClusterDescriptorAttributePartsListID)]];
+    if (!rootDescriptorPartsListDataValue || ![MTRArrayValueType isEqualToString:rootDescriptorPartsListDataValue[MTRTypeKey]]) {
+        return;
+    }
+    NSArray * partsList = rootDescriptorPartsListDataValue[MTRValueKey];
+    if (![partsList isKindOfClass:[NSArray class]] || !partsList.count) {
+        MTR_LOG_ERROR("%@ unexpected type %@ for parts list %@", self, [partsList class], partsList);
```

Comment:
So how would we get here?  Only if we mess up our own data structures, right?  Do we need these various guards?

Reply:
Guards are good :)

Reply:
Ever since the persistence PR, I've felt that we should add more guards around things. I am planning to add more as I touch code around MTRDevice, and also think about failure modes and if something extra needs to be done when issues are detected.